compiler-core/src/language_server/type_annotations.rs

```diff
@@ -0,0 +1,92 @@
+use super::code_action::CodeActionBuilder;
+use crate::{
+    ast::{TypedFunction, TypedModuleConstant},
+    line_numbers::LineNumbers,
+    type_::pretty::Printer,
+};
+use lsp_types::{
+    CodeActionKind, InlayHint, InlayHintKind, InlayHintLabel, Position, Range, TextEdit, Url,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct TypeAnnotations {
+    annotations: Vec<(Position, String)>,
+}
+
+impl TypeAnnotations {
+    pub fn from_module_constant(
+        constant: &TypedModuleConstant,
+        line_numbers: &LineNumbers,
+    ) -> Self {
+        if constant.annotation.is_none() {
+            let linecol = line_numbers.line_and_column_number(constant.location.end);
+            let position = Position::new(linecol.line - 1, linecol.column - 1);
+            let type_text = Printer::new().pretty_print(&constant.type_, 0);
+            let text = format!(": {type_text}");
+
+            Self {
+                annotations: vec![(position, text)],
+            }
+        } else {
+            Self::default()
+        }
+    }
+
+    pub fn from_function_definition(function: &TypedFunction, line_numbers: &LineNumbers) -> Self {
+        let mut annotations = vec![];
+        if function.return_annotation.is_none() {
+            let linecol = line_numbers.line_and_column_number(function.location.end);
+            let position = Position::new(linecol.line - 1, linecol.column - 1);
+            let type_text = Printer::new().pretty_print(&function.return_type, 0);
```

Comment:
I think this will generate invalid code when the type is imported in a qualified fashion, when there are type parameters and they have already been specified for some arguments, or when there are multiple types in the type with the same name.

Could we have tests for these please?

Seeing as we don't care about formatting etc I think it may be better to have a dedicated type printer for this use case rather than reusing the UI pretty printer. It'll be simpler than teaching the pretty printer about scopes, having to track type variable names, and having to conditionally disable the other features of the pretty printer.

Reply:
Could you give me some code examples of the cases you listed?

Reply:
```gleam
pub identity(value: some_type) {
  value
}
```
```gleam
```gleam
pub equal(v1: some_type, v2) {
  v1 == v2
}
```
```gleam
import mod1 as a
import mod2 as b

pub fn main() {
  let x: a.Value
  let y: b.Value
  #(x, y)
}
```
```gleam
import mod1.{Value as V1}
import mod2.{Value as V2}

pub fn main() {
  let x: V1
  let y: V2
  #(x, y)
}
```